### PR TITLE
feat: [UI - Wallet Stability] Create Max button component

### LINKED ISSUE #15066: [UI - Wallet Stability] `Max` button component
image

- create a reusable component (Swap and Send modal both have a slightly different copy atm)
- use a plain `StatusButton` instead of `StatusListItemTag`
- take an unformatted `max` value, `symbol` (for fiat values), `valid` and `clickable` bool flags
- format the max value with configurable decimal places and optional symbol
- should act as a button, emit `clicked()` signal
- provide a "safe" max output value, to be used for value (string) based input components
- integrate in storybook
- provide basic QML tests

Example usage:

image

### PR DESCRIPTION
**Pull request by caybro**
### What does the PR do

- create a reusable "Max" send button component
- use it in SwapModal and SendModal
- add some more tests in tst_SwapInputPanel.qml

Fixes #15066

### Affected areas

SwapModal, SendModal

### Screenshot of functionality (including design for comparison)

- [x] I've checked the design and this PR matches it

[Záznam obrazovky z 2024-06-10 12-41-32.webm](https://github.com/status-im/status-desktop/assets/5377645/57ab53af-acae-4289-bb21-2e18c3126a3d)

### CODE REVIEW

File: ui/imports/shared/popups/send/views/AmountToSend.qml [hunk 1]
```diff
@@ -127,10 +127,6 @@ ColumnLayout {
         Layout.fillWidth: true
         id: topItem
 
-        property double topAmountToSend: !inputIsFiat ? d.cryptoValueToSend
-                                                      : d.fiatValueToSend
-        property string topAmountSymbol: !inputIsFiat ? d.selectedSymbol
-                                                      : root.currentCurrency
```
Comment (caybro) [inline ui/imports/shared/popups/send/views/AmountToSend.qml:133]: unused code

File: ui/app/AppLayouts/Wallet/panels/SwapInputPanel.qml [hunk 2]
```diff
@@ -247,26 +240,26 @@ Control {
...
-                tagClickable: true
-                closeButtonVisible: false
-                titleText.font.pixelSize: 12
-                bgColor: amountToSendInput.input.valid || !amountToSendInput.input.text ? Theme.palette.primaryColor3 : Theme.palette.dangerColor2
-                titleText.color: amountToSendInput.input.valid || !amountToSendInput.input.text ? Theme.palette.primaryColor1 : Theme.palette.dangerColor1
-                onTagClicked: {
-                    const max = d.prepareForMaxSend(d.maxInputBalance, d.inputSymbol)
+
+                onClicked: {
+                    const max = maxSafeValue
```
Comment (Khushboo-dev-cpp) [inline ui/app/AppLayouts/Wallet/panels/SwapInputPanel.qml:260]: doesnt need to be created as const anymore? can be used directly?

**Review by Khushboo-dev-cpp — APPROVED** [review]
LGTM :)

File: ui/app/AppLayouts/Wallet/panels/SwapInputPanel.qml (hunk 2 above)
Comment (caybro) [reply ui/app/AppLayouts/Wallet/panels/SwapInputPanel.qml:260]:
```qml
const max = maxSafeValue
if (max > 0)
  amountToSendInput.input.text = maxSafeValueAsString
```

That's a numeric value used for the check; below we use the string/text value to insert into the input field

**Comment by caybro** [comment]
Tests will be fixed with https://github.com/status-im/status-desktop/pull/15116

File: ui/app/AppLayouts/Wallet/panels/SwapInputPanel.qml (hunk 2 above)
Comment (caybro) [reply ui/app/AppLayouts/Wallet/panels/SwapInputPanel.qml:260]: Oh you mean changing it to this:
```qml
if (maxSafeValue > 0)
```
?

Comment (caybro) [reply ui/app/AppLayouts/Wallet/panels/SwapInputPanel.qml]: Got it, ammended :)

**Review by micieslak — COMMENTED** [review]
LGTM in general, some points for consideration found.

And I'm not sure if the behavior here is intended:
[Screencast from 11.06.2024 10:01:12.webm](https://github.com/status-im/status-desktop/assets/20650004/ffac0825-48e4-468d-ae49-c768d803122e)

1. Should the value in the input be so long after clicking max?
2. The value still can be increased without warning.

File: ui/imports/shared/popups/send/SendModal.qml [hunk 3]
```diff
@@ -277,36 +271,23 @@ StatusDialog {
...
+
+                        value: d.maxInputBalance
+                        symbol: d.inputSymbol
+                        valid: amountToSendInput.input.valid || !amountToSendInput.input.text
+                        valueIsFiat: amountToSendInput.inputIsFiat
+                        formatCurrencyAmount: (amount, symbol, options) => d.currencyStore.formatCurrencyAmount(amount, symbol, options)
+
+                        onClicked: {
+                            if (maxSafeValue > 0)
+                                amountToSendInput.input.text = maxSafeValueAsString
```
[the remark this answers is not in the document]
Comment (caybro) [reply ui/imports/shared/popups/send/SendModal.qml:287]: Yeah, and the assumption is taken here in SendModal, not in any of those independent components (ie not a change really, it's been like that before)

File: ui/app/AppLayouts/Wallet/controls/MaxSendButton.qml [hunk 4]
```diff
@@ -0,0 +1,48 @@
...
+import StatusQ.Controls 0.1
+
+import utils 1.0
+
+StatusButton {
+    id: root
+
+    required property double value
+    required property string symbol
+    required property bool valid
```
[the remark this answers is not in the document]
Comment (caybro) [reply ui/app/AppLayouts/Wallet/controls/MaxSendButton.qml:14]: Yeah... I'll give it some thought :) I don't like the component to take `formattedValue` from outside, ppl will either forget or use some wrong stuff like :wink: 

```qml
`${value.toLocaleString(Qt.locale())} USD`
```

[the remark this answers is not in the document]
Comment (micieslak) [reply ui/app/AppLayouts/Wallet/controls/MaxSendButton.qml:14]: People may forget setting this property, similarly as other properties. But it closes component for modifications, user has full freedom how to format without modifying the component. Finally we should end up with only few formatters on top of objects hierarchy passed to one to another subcomponent. It gives possibility to change formatting behavior easily from single (or fiew) place(s).

File: ui/app/AppLayouts/Wallet/controls/MaxSendButton.qml [hunk 5]
```diff
@@ -0,0 +1,48 @@
...
+    required property bool valueIsFiat
+    property var formatCurrencyAmount: (amount, symbol, options) => { return "FIXME" }
```
[the remark this answers is not in the document]
Comment (caybro) [reply ui/app/AppLayouts/Wallet/controls/MaxSendButton.qml]: Fixed :)

File: ui/app/AppLayouts/Wallet/controls/MaxSendButton.qml (hunk 4 above)
[the remark this answers is not in the document]
Comment (caybro) [reply ui/app/AppLayouts/Wallet/controls/MaxSendButton.qml:14]: > People may forget setting this property, similarly as other properties. 

No, I intentionally make all the necessary properties `required` but sadly we can't do this effectively with `property var someFunction: () => {}` 

I'll try to think of something in C++, got some ideas; meanwhile I think it's actually better to use sth like:
```qml
required property var formatCurrencyAmount //: (amount, symbol) => { return "IMPLEMENT ME" }
```

than betting on ppl to provide the correctly `formattedValue` from the outside, I've seen it hundreds of times, and ppl will rarely get it correct

[the remark this answers is not in the document]
Comment (caybro) [reply ui/app/AppLayouts/Wallet/controls/MaxSendButton.qml:14]: > ```qml
> property string formattedValue: `${value.toLocaleString(Qt.locale())} USD`
> ```

Another argument: if you provide the default implementation (e.g. as above) then there's no incentive to override it

**Comment by caybro** [comment]
> LGTM in general, some points for consideration found.
> 
> And I'm not sure if the behavior here is intended:

> 1. Should the value in the input be so long after clicking max?

Yup, we paste the literal (max safe) value (minus the ETH gas fees) in order to get just under the max value; so instead of rounding, we rather truncate the value for display, and paste the full one

> 2. The value still can be increased without warning.

Yes, but still within the max limits; question then is, should we display the rounded value or the truncated value as part of the button's text? CC @benjthayer 

E.g. for "2.765", should we display "2.77" (rounding) or "2.76" (truncating)? If we display "2.77" and then paste that into the input field, we immediately get ourselves caught in an error state

**Comment by Khushboo-dev-cpp** [comment]
@caybro hope you are coverring this https://www.figma.com/design/TS0eQX9dAZXqZtELiwKIoK/Swap---Milestone-1?node-id=3901-187853&m=dev, we show `Max:0` instead of `No Balances Active`

**Comment by caybro** [comment]
> @caybro hope you are coverring this https://www.figma.com/design/TS0eQX9dAZXqZtELiwKIoK/Swap---Milestone-1?node-id=3901-187853&m=dev, we show `Max:0` instead of `No Balances Active`

Ah good point, will amend!

**Comment by caybro** [comment]
> @caybro hope you are coverring this https://www.figma.com/design/TS0eQX9dAZXqZtELiwKIoK/Swap---Milestone-1?node-id=3901-187853&m=dev, we show `Max:0` instead of `No Balances Active`

Fixed :)

image

**Comment by benjthayer** [comment]
> > LGTM in general, some points for consideration found.
> > And I'm not sure if the behavior here is intended:
> 
> > 1. Should the value in the input be so long after clicking max?
> 
> Yup, we paste the literal (max safe) value (minus the ETH gas fees) in order to get just under the max value; so instead of rounding, we rather truncate the value for display, and paste the full one
> 
> > 2. The value still can be increased without warning.
> 
> Yes, but still within the max limits; question then is, should we display the rounded value or the truncated value as part of the button's text? CC @benjthayer
> 
> E.g. for "2.765", should we display "2.77" (rounding) or "2.76" (truncating)? If we display "2.77" and then paste that into the input field, we immediately get ourselves caught in an error state

I think it makes sense to truncate instead of round in this instance - as you say, rounding may lead to validation which would be odd.

File: ui/app/AppLayouts/Wallet/WalletUtils.qml [hunk 6]
```diff
@@ -57,4 +57,13 @@ QtObject {
         }
         return hovered? WalletUtils.colorizedChainPrefix(chainShortNames) + Utils.richColorText(finalAddress, Theme.palette.directColor1) : chainShortNames + finalAddress
     }
+
+    // compensate for the amount of eth to pay for the gas
+    function prepareForMaxSend(value, symbol) {
```
Comment (micieslak) [inline ui/app/AppLayouts/Wallet/WalletUtils.qml:62]: maybe `calculateMaxSafeSendAmount`? `prepareForMaxSend` is quite vague, especially outside context of max button.
I would also add explanation that `value` is maximum value intended to use for the transaction, usually total balance for given token.

File: ui/app/AppLayouts/Wallet/controls/MaxSendButton.qml (hunk 4 above)
[the remark this answers is not in the document]
Comment (micieslak) [reply ui/app/AppLayouts/Wallet/controls/MaxSendButton.qml:16]: Ok, fair enough :)

File: ui/app/AppLayouts/Wallet/WalletUtils.qml (hunk 6 above)
Comment (caybro) [reply ui/app/AppLayouts/Wallet/WalletUtils.qml:62]: Alright

File: ui/app/AppLayouts/Wallet/controls/MaxSendButton.qml [hunk 7]
```diff
@@ -0,0 +1,41 @@
...
+
+StatusButton {
+    id: root
+
+    required property double value
+    required property string symbol
+    required property bool valid
+    property var formatCurrencyAmount: (amount, symbol) => { return "FIXME" }
+
+    readonly property double maxSafeValue: WalletUtils.prepareForMaxSend(value, symbol)
```
Comment (micieslak) [inline ui/app/AppLayouts/Wallet/controls/MaxSendButton.qml:19]: Ideally this value should be provided from outside, then there is no need to make call to `WalletUtils.prepareForMaxSend(value, symbol)` internally.

**Review by micieslak — APPROVED** [review]

File: ui/app/AppLayouts/Wallet/controls/MaxSendButton.qml (hunk 7 above)
Comment (caybro) [reply ui/app/AppLayouts/Wallet/controls/MaxSendButton.qml]: I'd really keep that call inside (as it always was); even that function is used only here and nowhere else

**Pull request merged**